File: 프로그래머스/3/12979. 기지국 설치/기지국 설치.js
function solution(n, stations, w) {
    let s = 0;
    let [now, result] = [1, 0];
    while (1) {
        if (now > n) break;
        if (s < stations.length) {
            const front = stations[s];
            if (now >= front - w && now <= front + w) {
                now = stations[s++] + w + 1;
                continue;
            }
        } else {
            if (now + w > n) {
                result += 1;
                break;
            }
        }
        now += w * 2 + 1;
        result += 1;
    }
    return result;
}

File: 프로그래머스/3/12979. 기지국 설치/기지국 설치.py
import math

def solution(n, stations, w):
    answer = 0
    cover = 2 * w + 1
    start = 1

    for s in stations:
        left = s - w
        right = s + w

        if start < left:
            gap = left - start
            answer += math.ceil(gap / cover)

        start = right + 1

    if start <= n:
        gap = n - start + 1
        answer += math.ceil(gap / cover)

    return answer